sklearn/feature_selection/mutual_info_.py

```diff
@@ -247,8 +247,11 @@ def _estimate_mi(X, y, discrete_features='auto', discrete_target=False,
     X, y = check_X_y(X, y, accept_sparse='csc', y_numeric=not discrete_target)
     n_samples, n_features = X.shape
 
-    if discrete_features == 'auto':
-        discrete_features = issparse(X)
+    if isinstance(discrete_features, str):
```

Comment:
Maybe have this whole block:

```
    if isinstance(discrete_features, str):
        if discrete_features == 'auto':
            discrete_features = issparse(X)
        else:
            raise ValueError("Invalid string value for discrete_features.")

    if isinstance(discrete_features, bool):
        discrete_mask = np.empty(n_features, dtype=bool)
        discrete_mask.fill(discrete_features)
    else:
        discrete_features = np.asarray(discrete_features)
        if discrete_features.dtype != 'bool':
            discrete_mask = np.zeros(n_features, dtype=bool)
            discrete_mask[discrete_features] = True
        else:
            discrete_mask = discrete_features
```

in a series of if/elses and also use `utils.validaton.check_array` instead of
`np.asarray`?

Then in the last `else` you can raise an exception and complain about the given
`discrete_features` not supported.

Reply:
@adrinjalali if I'm understanding you correctly, in order to change this to a series of if/elifs with a final else where exception is raised then will need to explicitly test if `discrete_features` is "array like" instead of letting this case fall to final else.  Currently numpy.asarray conversion is implicitly doing that.  For second elif so there a better way to do that?

```python
if isinstance(discrete_features, str):
    ...
elif isinstance(discrete_features, bool):
    ...
elif check_array(discrete_features):
    discrete_features = check_array(discrete_features)
    ...
else:
   raise ValueError("Invalid value for discrete_features.")

```

Reply:
Yes, I think it's a good idea to explicitly check if it's array-like. Just note that `check_array` returns the array, not a boolean. Maybe something like:
``` python
if isinstance(discrete_features, str):
    ...
elif isinstance(discrete_features, bool):
    ...
else:
    try:
        discrete_features = check_array(discrete_features)
        ...
    except:
       raise ValueError("Invalid value for discrete_features.")

Reply:
> ```python
> if isinstance(discrete_features, str):
>     ...
> elif isinstance(discrete_features, bool):
>     ...
> else:
>     try:
>         discrete_features = check_array(discrete_features)
>         ...
>     except:
>        raise ValueError("Invalid value for discrete_features.")
> ```

Don't think we need the try/except since `check_array` already raises appropriate errors.  In latest commit I finished if statement impl.